### DISCUSSION TITLE
How to convert an anonymous user to a permanent user with their password?

### DISCUSSION THREAD

**Discussion by VValiente** [supabase/supabase#29017] [Questions]
Hello, how should we convert an anonymous user to a permanent user with their password? 

At app launch, if no user is found, the app signs the user anonymously and I'm saving some onboarding data under their anonymous id. Then at the end part of the onboarding, the app asks them to create an account. If they choose to create an account with email and password, what should be the proper steps to link the email and password?

Calling 
```
let user = try await supabase.auth.update(user: UserAttributes(email: email, password: password), redirectTo: nil)
```
throws the following error:
```
api(Auth.AuthError.APIError(msg: Optional("Updating password of an anonymous user is not possible"), code: Optional(422), error: nil, errorDescription: nil, weakPassword: nil))
```

I tried to first update just the email but the `user.isAnonymous` remains `true`. Please advise, because I can't find this info in the documentation about converting anonymous users.

I'm on Supabase 2.17.0 for iOS (https://github.com/supabase/supabase-swift.git)

**Comment by GaryAustin1** [comment]
https://supabase.com/docs/guides/auth/auth-anonymous?queryGroups=language&language=js#convert-an-anonymous-user-to-a-permanent-user

**Reply by VValiente** [reply]
I'm sorry but what exactly are you suggesting? That's the page I was following. I meant, I couldn't find the info to link an anonymous user with an email AND password. Sorry for the confusion.

The iOS Supabase version I am on has the extra parameter of `redirectTo:`. I'm not sure if I have to supply a value for it to make it work. I assume not since it's optional. 

I've also tried this:
1. Enable Allow Manual Linking for the project
2. In the app, the user is initially signed in anonymously if no user is found.
3. On the create account step, when signing up with email and password, I tried to link an identity(email) first as instructed in the documentation:
```
try await supabase.auth.linkIdentity(provider: .email, scopes: nil, redirectTo: nil, queryParams: [], launchURL: { _ in })
```
Please note that in the version of Supabase I'm using, there are these extra parameters. I do not know what is required to make this work, because manual linking always fails:
```
api(Auth.AuthError.APIError(msg: Optional("Unsupported provider: Provider email could not be found"), code: Optional(400), error: nil, errorDescription: nil, weakPassword: nil))
```

If I skip the linking identity above and jump straight to this:
```
let user = try await supabase.auth.update(user: UserAttributes(email: email), redirectTo: nil)
let identities = user.identities
```
It succeeds but the `identities` remain empty, and in the Auth table, the user is still Anonymous.

If I do it in this order:
```
var user = try await supabase.auth.update(user: UserAttributes(email: email), redirectTo: nil)
try await supabase.auth.linkIdentity(provider: .email, scopes: nil, redirectTo: nil, queryParams: [], launchURL: { _ in })
let identities = user.identities
```
It fails on `linkIdentity` with the same error:
```
api(Auth.AuthError.APIError(msg: Optional("Unsupported provider: Provider email could not be found"), code: Optional(400), error: nil, errorDescription: nil, weakPassword: nil))
```

Is the manual linking not yet working? It is marked as beta. Or did I miss a step? I'm not using other providers (e.g. `.google`, `.apple`) at this point. I wanted to make the email+password to work first.

**Reply by silentworks** [reply]
The user should be signed in anonymously and then you call the `supabase.auth.update` function and pass the users email. If email confirmation is turned on then the identity won't be updated right away until the link in the email that was sent by Supabase auth for Email change has been clicked. I've attached a video showing this, albeit not the Swift client but it should be the same flow.

https://github.com/user-attachments/assets/64b0ddcc-c70f-4fe1-b89a-25111378dca7

**Reply by VValiente** [reply]
@silentworks thanks for testing it on your side.

The user is signed anonymously at the very start of the app if no logged-in user is found. Then at the end part of the onboarding, they are asked to create an account. This is where I was trying to do the manual linking. I had turned off the email confirmation requirement and as I wrote in detail in my previous post, the provider remains Anonymous, unlike in your video. 

Since I want the user to create an account (meaning with a password), I ditched anonymous sign-in and just saved their onboarding state locally.
